pkg/cache/resource_node.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"maps"
+
+	"sigs.k8s.io/kueue/pkg/resources"
+)
+
+type ResourceNode struct {
+	// Quotas are the ResourceQuotas specified for the current
+	// node.
+	Quotas map[resources.FlavorResource]*ResourceQuota
+	// SubtreeQuota is the sum of the node's quota, as well as
+	// resources available from its children, constrained by
+	// LendingLimits.
+	SubtreeQuota resources.FlavorResourceQuantities
+	// Usage is the quantity which counts against this node's
+	// SubtreeQuota. For ClusterQueues, this is simply its
+	// usage. For Cohorts, this is the sum of child usage past
+	// guaranteedQuotas.
+	Usage resources.FlavorResourceQuantities
+}
+
+func NewResourceNode() ResourceNode {
+	return ResourceNode{
+		Quotas:       make(map[resources.FlavorResource]*ResourceQuota),
+		SubtreeQuota: make(resources.FlavorResourceQuantities),
+		Usage:        make(resources.FlavorResourceQuantities),
+	}
+}
+
+func (r ResourceNode) Clone() ResourceNode {
+	return ResourceNode{
+		Quotas:       maps.Clone(r.Quotas),
+		SubtreeQuota: maps.Clone(r.SubtreeQuota),
+		Usage:        maps.Clone(r.Usage),
+	}
+}
+
+// guaranteedQuota is the capacity which will not be lent the node's
+// Cohort.
+func (r ResourceNode) guaranteedQuota(fr resources.FlavorResource) int64 {
+	if lendingLimit := r.Quotas[fr].lendingLimitSafe(); lendingLimit != nil {
+		return max(0, r.SubtreeQuota[fr]-*lendingLimit)
+	}
+	return 0
+}
+
+type wiredResourceNode[T any] interface {
```

Comment:
if you can't apply any of my other remarks, then this should be renamed to `hierarchicalResourceNode`. But I hope we can get rid of this interface.